policy: Add GetAuthTypes()

jrajahalme opened the pull request:
Track the required authentication mode for all selected remote identities
in a selector policy, and add GetAuthTypes() that can be used to find out
the required authentication modes between a local and remote security
identity.

We recently added authentication requirement support to Cilium Network Policy. This is currently plumbed to bpf datapath policymap values so that the datapath may enforce authentication status of traffic by consulting the new bpf authmap. Authmap is updated by Cilium Agent's auth manager (`pkg/auth`), currently based on the "authentication required" signals delivered from the datapath to the agent via signalmap. Once authentication is done, auth manager updates authmap and datapath starts accepting traffic.

The new functionality provided in this PR is needed for the auth manager to be able to decide if an expiring authentication needs to be renewed (re-authenticated). Without this functionality re-authentication between given Cilium security IDs (local/remote) will be done perpetually, or until either the security ID is removed, or the node with which the authentication was done, is removed. The new `GetAuthTypes()` function is intended to be called by the auth manager to potentially end re-authentication for the given local/remote security ID pair, when no AuthTypes are returned, indicating that the policy has changed to not require authentication between the given local/remote security ID any more. 

https://github.com/cilium/cilium/pull/26068 uses this new functionality and is currently blocked for this functionality to be added.

jrajahalme committed:
policy: Cache selector policies in distillery test

Use PolicyCache in unit tests.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add GetAuthType()

Track the required authentication mode for all selected remote identities
in a selector policy, and add GetAuthType() that can be used to find out
the required authentication mode between a local and remote security
identity.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter commented on pkg/policy/distillery.go:
```diff
@@ -153,6 +153,34 @@ func (cache *PolicyCache) UpdatePolicy(identity *identityPkg.Identity) error {
+// GetAuthType returns the AuthType required by the policy between the localID and remoteID, or
+// AuthTypeDisabled if none.
+//
+// Note that the first found AuthType is returned. This is correct as the policy should only
```
technically, it's possible to have multiple auth types for different protocols & ports between the same identities.

so i wonder whether it would be ok to just return whether auth is enabled between the two identities?

mhofstetter replied:
this has been "fixed" with the latest changes! going to resolve.

mhofstetter reviewed (approved):
LGTM - thanks @jrajahalme 

only the small non-blocking question regarding the API (first matching auth type vs bool)

jrajahalme committed:
policy: Track and return multiple AuthTypes for GetAuthTypes()

Track and return multiple AuthTypes for GetAuthTypes(), add multiple auth
types to the distillery unit test.

While we currently plan to only support Spire AuthType, the policy engine
is wired for supporting multiple auth types. Honor this in the new
GetAuthTypes() internal API.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joestringer commented on pkg/policy/l4.go:
```diff
@@ -846,9 +852,23 @@ func (l4 *L4Filter) attach(ctx PolicyContext, l4Policy *L4Policy) {
+
+				if authType := l7policy.GetAuthType(); authType != AuthTypeDisabled {
+					if l4Policy.AuthMap == nil {
+						l4Policy.AuthMap = make(AuthMap, 1)
+					}
+					authTypes := l4Policy.AuthMap[cs]
+					if authTypes == nil {
+						authTypes = make(AuthTypes, 1)
```
nit: Why `1`? Presumably these should be sized based on the number of authentication rules for any particular endpoint (well, local endpoint identity) instead?

Background I'm thinking here is that maybe we are trying to set the default map size in order to micro-optimize for memory allocation, but we end up shooting ourselves in the foot because we force reallocations if there's more than one authenticated peer for the endpoint?

I guess overall you might argue that overall "number of unique peers an app connects to" amortizes towards one peer identity and hence selected by one rule :thinking:.

A cursory search around Go default map capacity seems to suggest one bucket, eight entries... do we think there's a strong reason to be particular about this aspect here?

joestringer replied:
I'm thinking worst case we could end up recalculating new selectorPolicy on a regular basis, and every time we first allocate a map that's too small, then allocate again to a size that fits the policy. Mildly wasteful.

Anyway this is not a big deal, just seems like it may be premature optimization.

joestringer commented on pkg/policy/distillery.go:
```diff
@@ -153,6 +153,45 @@ func (cache *PolicyCache) UpdatePolicy(identity *identityPkg.Identity) error {
 	return err
 }
 
+// GetAuthTypes returns the AuthTypes required by the policy between the localID and remoteID, if
+// any, otherwise returns nil.
```
nit: `ID` can end up ambiguous, it'd be nice to give the reader one less thing to be confused about - stating clearly that the inputs are expected to be security identities. If by any weird course we end up plugging the wrong id numbers in, it could help point out the mistake.
```suggestion
// GetAuthTypes returns the AuthTypes required by the policy between the
// security identities localID and remoteID, if any, otherwise returns nil.
```

joestringer reviewed (approved):
Makes sense, just some super minor nits on a couple of style/docs aspects. Feel free to roll into the PR, submit separately, or disagree with me in the threads below :)
